server/cabal.project.freeze

```diff
@@ -0,0 +1,349 @@
+             any.postgresql-libpq ==0.9.4.2,
+             postgresql-libpq -use-pkg-config,
+             any.pretty ==1.1.3.6,
+             any.pretty-show ==1.9.5,
+             any.prettyprinter ==1.2.1,
+             prettyprinter -buildreadme,
+             any.primitive ==0.6.4.0,
+             any.primitive-extras ==0.7.1,
+             any.process ==1.6.5.0,
+             any.profunctors ==5.3,
+             any.protolude ==0.2.3,
+             any.psqueues ==0.2.7.1,
+             any.quickcheck-instances ==0.3.19,
+             quickcheck-instances -bytestring-builder,
+             any.quickcheck-io ==0.2.0,
+             any.random ==1.1,
+             any.reflection ==2.1.4,
+             reflection -slow +template-haskell,
+             any.regex-base ==0.93.2,
+             regex-base +newbase +splitbase,
+             any.regex-tdfa ==1.2.3.1,
+             regex-tdfa -devel,
+             any.reroute ==0.5.0.0,
+             any.resource-pool ==0.2.3.2,
+             resource-pool -developer,
+             any.resourcet ==1.2.2,
+             any.retry ==0.7.7.0,
+             retry -lib-werror,
+             any.rts ==1.0,
+             any.safe ==0.3.17,
+             any.scientific ==0.3.6.2,
+             scientific -bytestring-builder -integer-simple,
+             any.semigroupoids ==5.3.2,
+             semigroupoids +comonad +containers +contravariant +distributive +doctests +tagged +unordered-containers,
+             any.semigroups ==0.18.5,
+             semigroups +binary +bytestring -bytestring-builder +containers +deepseq +hashable +tagged +text +transformers +unordered-containers,
+             any.semver ==0.3.3.1,
+             any.setenv ==0.1.1.3,
+             any.shakespeare ==2.0.22,
+             shakespeare -test_coffee -test_export -test_roy,
+             any.simple-sendfile ==0.2.28,
+             simple-sendfile +allow-bsd,
+             any.socks ==0.5.6,
+             any.split ==0.2.3.3,
+             any.statistics ==0.15.0.0,
+             any.stm ==2.5.0.0,
+             any.stm-containers ==1.1.0.4,
+             any.stm-hamt ==1.2.0.2,
+             any.streaming-commons ==0.2.1.0,
+             streaming-commons -use-bytestring-builder,
+             any.string-conversions ==0.4.0.1,
+             any.superbuffer ==0.3.1.1,
+             any.tagged ==0.8.6,
+             tagged +deepseq +transformers,
+             any.template-haskell ==2.14.0.0,
+             any.terminal-size ==0.3.2.1,
+             any.text ==1.2.3.1,
+             any.text-builder ==0.6.5,
+             any.text-conversions ==0.3.0,
+             any.text-latin1 ==0.3.1,
+             any.text-printer ==0.5,
+             any.text-short ==0.1.2,
+             text-short -asserts,
+             any.tf-random ==0.5,
+             any.th-abstraction ==0.2.11.0,
+             any.th-lift ==0.7.11,
+             any.th-lift-instances ==0.1.12,
+             any.these ==0.7.6,
+             any.time ==1.8.0.2,
```

Comment:
Probably off-topic, say I need time = 1.9.3 but that has other deps , then how will i figure that out? `cabal-install` will resolve the plan and suggest changes?

Reply:
The easiest way is just to delete the `caba.project.freeze` file an re-run `cabal new-freeze`. The main point of the freeze file is to ensure reproducible builds, so other people won’t get spurious build failures because the state of the package index changed and for some reason the new build plan is somehow broken. It’s to avoid things from changing out from underneath you. There’s nothing wrong with just blowing it away and regenerating it, since once you check it in, you’ve verified that the build plan is, in fact, good.

Stackage is really the same way. Technically it’s true that bumping the LTS version only bumps minor versions, so theoretically it’s “safer” if people are versioning their packages really carefully, but the truth is that Stackage does no auditing of packages to ensure that’s actually true; all that matters about an LTS is that all the packages build together. When Stackage was first created, this was a bigger deal than it is now, since a lot of packages *didn’t* build together, so `cabal-install` often failed to generate build plans. But the funny thing about Stackage is that its benefits are actually useful to `cabal-install` users almost as much as to `stack` users, since the ecosystem-wide CI means packages are kept more compatible, and `cabal-install` is able to generate good build plans.

In that light, I feel like there are really relatively few benefits to using Stackage directly, in a “tragedy of the commons” sort of way. `cabal-install` is more flexible, since you aren’t tethered to a particular snapshot, but you still get working build plans without much effort.

Reply:
> so other people won’t get spurious build failures because the state of the package index changed

Just for the record (please just ignore me if this is already known); if the main worry is the package-index changing resulting in unsound build-plans, "freezing" merely the index-state can be accomplished by `--index-state=<timestamp>` instead of freezing the actual build-plan.

But of course, doing a full `cabal freeze` via version pinning (and you'd still want to combine that with index-freezing) gives more guarantees  re reproducibility.

Reply:
@hvr 

>  (and you'd still want to combine that with index-freezing) 

Why freeze the index as well? In case a hackage "revision" changes the constraints for a dependency version retroactively?

@lexi-lambda I'd still like a little clarity on how we want to manage package version upgrades and new dependencies. I think it might make sense to pin the `index-state` in `cabal.project` (similar to a stackage version) if it allows us to do two things separately:

- add new dependencies without upgrading a bunch of others
- upgrade some or all dependencies to newer versions (and bumping index version)

...but I'm not sure that's how that would work.

Reply:
Maybe I’m in the minority on this, but I’ve always been of the opinion that the hand-wringing about Hackage revisions is largely overblown. I understand why a project like Stackage might find them frustrating, but Stackage has pretty special needs. From my point of view, Hackage revisions are an extremely good thing, as they prevent incorrectly-specified constraints from indefinitely polluting the solver.

My guess is that we won’t ever encounter a situation where a revision causes a build that previously passed to fail. Revisions are almost exclusively added for one of two reasons:

  1. A new version of a dependency was released, so someone makes a revision to weaken a conservative upper bound once it’s clear the new version is still compatible.

  2. Someone discovered that a selected build plan does not actually successfully compile, so they fix the incorrect bound via revision.

We don’t care about (1), since we’ve pinned all the versions in the freeze file anyway. Weakened bounds can’t affect our build plans. And while it’s theoretically possible that we could get bitten by (2), it’s highly unlikely, as if our build passes CI, then our build plan fundamentally must be a good one. If a package maintainer adds a revision that breaks our previously-working build plan, either they screwed up (and we should tell them about it), or they had a really good reason to do so that we probably want to know about!

Given the above, I don’t think pinning the index state is super useful, and it seems more likely to create busywork than anything else. If it ever becomes a problem, I’m more than happy to revisit this decision, but I don’t expect it to be.

> I'd still like a little clarity on how we want to manage package version upgrades and new dependencies as a team.

From my point of view, the right workflow here is really simple: delete the freeze file and re-run `new-freeze`. Ideally, this should always be possible, as if it isn’t, that means the constraints in our `.cabal` file are too loose! The purpose of the freeze file in my eyes is simply build reproducibility: if it turns out the constraints in our `.cabal` file *are* too loose, we don’t want to find out because CI jobs just start suddenly failing. But if we find out because we deleted the freeze file, that’s great, as there’s a human in the loop at that point to fix the `.cabal` file.

In other words, I think manually worrying about the freeze file is an antipattern. We should never rely on the constraints in the freeze file to save us, so we should only be concerned with the `.cabal` file. That said, if you *really* want to bump just one dependency, you can either:

  1. Update the constraint in the freeze file by hand.

  2. Delete the constraint for that particular package from the freeze file and run `new-freeze` again.

     This should work, as I believe `cabal-install` will take the existing freeze file into account when creating the new build plan. When it regenerates the freeze file, it will include all the existing constraints, plus whatever new ones are needed for the updated dependency.

But really, I think that’s overly pessimistic. If we generate a build plan that (a) compiles and (b) passes CI, then we can be pretty confident it’s a good one. And if anything, we *want* our dependencies to be as up to date as possible, since we want to integrate bugfix patches as much as we can!

Let me know if that makes sense to you or if you disagree with any of the above. I can also put some of these thoughts about the workflow in the CONTRIBUTING.md if you think that would help.

Reply:
> Hackage revisions are an extremely good thing,

I agree, I'm not really worried about revisions.

> That said, if you really want to bump just one dependency, you can either: ...

I have pretty high confidence in the quality of libraries on hackage generally, but if we have a situation where e.g. many dependencies are changing arbitrarily (maybe even being downgraded?) without much oversight on each PR that sounds a little scary for debugging issues (esp performance issues). 

I'm okay trying to proceed this way. But I think I'd be more comfortable with guidance like: "try to upgrade dependencies only as needed by editing the freeze file ... upgrades of many dependencies can happen, but should be in a separate commit so it can be easily reverted if necessary". Does that sound reasonable?

Reply:
I don’t expect there to be a ton of churn in the freeze file even if people delete it and recreate it, but I’m willing to try and be conservative if possible if you think that’s a good idea.

From my point of view, there is a fundamental tension here: we want to incorporate bugfix patches as much as possible (perhaps there’s a performance bug or security vulnerability), and we just aren’t going to regularly audit all of our dependencies to see which ones need upgrading. For that reason, my preference is to give package authors the benefit of the doubt and to assume that newer versions are better. Sure, we could get bitten by upgrading if that assumption turns out to be wrong, but we could also get bitten if a package fixes an issue and we haven’t upgraded.

Still, I can also understand wanting to avoid profiling noise that comes from changes in dependencies rather than changes in our code. So I’m willing to be a little more deliberate about this, I just don’t want to be more conservative than we have to be.